### PR TITLE
cli: rados: support for high precision time using stat2

### PR DESCRIPTION
**Pull request by theanalyst**
This commit introduces `stat2` option for the rados cli, which is
similar to `stat` except that it returns the mtime in high precision,
which is useful for inspecting objects, for example, if the application
had used the related librados api calls (like radosgw using `mtime2`)

Fixes: http://tracker.ceph.com/issues/21199
Signed-off-by: Abhishek Lekshmanan <abhishek@suse.com>

### CODE REVIEW

File: src/tools/rados/rados.cc [hunk 1]
```diff
@@ -2253,6 +2254,30 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
+      usage_exit();
+    string oid(nargs[1]);
+    uint64_t size;
+    struct timespec mtime;
+    if (use_striper) {
+      ret = striper.stat2(oid, &size, &mtime);
+    } else {
+      ret = io_ctx.stat2(oid, &size, &mtime);
+    }
+
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: please drop the empty line.

File: src/tools/rados/rados.cc [hunk 2]
```diff
@@ -2253,6 +2254,30 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
+    if (ret < 0) {
+      cerr << " error stat-ing " << pool_name << "/" << oid << ": "
+	   << cpp_strerror(ret) << std::endl;
+      goto out;
+    } else {
+      utime_t t(mtime);
+      cout << pool_name << "/" << oid
+	   << " mtime " << t << ", size " << size << std::endl;
+    }
+
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: ditto.

File: src/tools/rados/rados.cc [hunk 3]
```diff
@@ -2253,6 +2254,30 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
+  }
+
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: ditto.

File: src/tools/rados/rados.cc (hunk 1 above)
Comment (theanalyst) [reply src/tools/rados/rados.cc]: ack fixing

File: doc/man/8/rados.rst [hunk 4]
```diff
@@ -60,7 +60,8 @@ Options
 .. option:: --striper
 
    Uses the striping API of rados rather than the default one.
-   Available for stat, get, put, append, truncate, rm, ls and all xattr related operation
+   Available for stat, stat2, stat2, get, put, append, truncate, rm, ls
```
Comment (tchaikov) [inline doc/man/8/rados.rst]: dedup stat2, please.

Comment (theanalyst) [reply doc/man/8/rados.rst]: oh copy-paste error :) fixing

File: src/tools/rados/rados.cc [hunk 5]
```diff
@@ -94,6 +94,7 @@ void usage(ostream& out)
 "   setxattr <obj-name> attr val\n"
 "   rmxattr <obj-name> attr\n"
 "   stat <obj-name>                  stat the named object\n"
+"   stat2 <obj-name>                 stat2 the named object (with high precision time)\n"
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: the prefix should be simply `rados` or `tools/rados`.

Comment (theanalyst) [reply src/tools/rados/rados.cc]: for the commit message right? fixing

Comment (tchaikov) [reply src/tools/rados/rados.cc]: yeah =)

**Comment by xiexingguo** [comment]
http://pulpito.ceph.com/xxg-2017-09-01_07:17:41-rados-wip-xxg-testing-2017-09-01-1357-distro-basic-smithi/
http://pulpito.ceph.com/xxg-2017-09-03_12:51:11-rados-wip-xxg-testing-2017-09-01-1357-distro-basic-smithi/

**Pull request merged**